File: Numerical differentiation/numerical_diff.cpp
#include "numerical_diff.hpp"

using namespace std;

TestFunction::TestFunction(
    const string& name,
    MathFunction function,
    MathFunction exactDerivative
) {
    this->name = name;
    this->function = function;
    this->exactDerivative = exactDerivative;
}

string TestFunction::getName() const {
    return this->name;
}

double TestFunction::evaluate(double x) const {
    return this->function(x);
}

double TestFunction::exact(double x) const {
    return this->exactDerivative(x);
}

DifferenceMethod::DifferenceMethod(const string& name) {
    this->methodName = name;
}

string DifferenceMethod::getName() const {
    return this->methodName;
}

// ---------------- ForwardDifference ----------------

ForwardDifference::ForwardDifference()
    : DifferenceMethod("Forward") {}

double ForwardDifference::derivative(
    const TestFunction& f, double x, double h
) const {
    return (f.evaluate(x + h) - f.evaluate(x)) / h;
}

// ---------------- BackwardDifference ----------------

BackwardDifference::BackwardDifference()
    : DifferenceMethod("Backward") {}

double BackwardDifference::derivative(
    const TestFunction& f, double x, double h
) const {
    return (f.evaluate(x) - f.evaluate(x - h)) / h;
}

// ---------------- CentralDifference ----------------

CentralDifference::CentralDifference()
    : DifferenceMethod("Central") {}

double CentralDifference::derivative(
    const TestFunction& f, double x, double h
) const {
    return (f.evaluate(x + h) - f.evaluate(x - h)) / (2.0 * h);
}

File: Python_Matrix/Numerical_differentiation/numerical_diff.py
from abc import ABC, abstractmethod


class TestFunction:
    def __init__(self, name, function, exact_derivative):
        self.name = name
        self.function = function
        self.exact_derivative = exact_derivative

    def get_name(self):
        return self.name

    def evaluate(self, x):
        return self.function(x)

    def exact(self, x):
        return self.exact_derivative(x)


class DifferenceMethod(ABC):
    def __init__(self, name):
        self.name = name

    def get_name(self):
        return self.name

    @abstractmethod
    def derivative(self, f, x, h):
        pass


class ForwardDifference(DifferenceMethod):
    def __init__(self):
        super().__init__("Forward")

    def derivative(self, f, x, h):
        return (f.evaluate(x + h) - f.evaluate(x)) / h


class BackwardDifference(DifferenceMethod):
    def __init__(self):
        super().__init__("Backward")

    def derivative(self, f, x, h):
        return (f.evaluate(x) - f.evaluate(x - h)) / h


class CentralDifference(DifferenceMethod):
    def __init__(self):
        super().__init__("Central")

    def derivative(self, f, x, h):
        return (f.evaluate(x + h) - f.evaluate(x - h)) / (2 * h)


class RichardsonExtrapolation(DifferenceMethod):
    def __init__(self):
        super().__init__("Richardson")
        self.central = CentralDifference()

    def derivative(self, f, x, h):
        d_h = self.central.derivative(f, x, h)
        d_h2 = self.central.derivative(f, x, h / 2)
        return (4 * d_h2 - d_h) / 3
